ledger/complete/compactor.go

```diff
@@ -0,0 +1,393 @@
+			// observer.OnComplete() is called when Compactor starts shutting down,
+			// which may close channel that observer.OnNext() uses to send data.
+			select {
+			case <-ctx.Done():
+				return nil
+			default:
+				observer.OnNext(checkpointNum)
+			}
+		}
+	}
+
+	return nil
+}
+
+func createCheckpoint(checkpointer *realWAL.Checkpointer, logger zerolog.Logger, tries []*trie.MTrie, checkpointNum int) error {
+
+	logger.Info().Msgf("serializing checkpoint %d", checkpointNum)
+
+	startTime := time.Now()
+
+	writer, err := checkpointer.CheckpointWriter(checkpointNum)
+	if err != nil {
+		return fmt.Errorf("cannot generate checkpoint writer: %w", err)
+	}
+	defer func() {
+		closeErr := writer.Close()
+		// Return close error if there isn't any prior error to return.
+		if err == nil {
+			err = closeErr
+		}
+	}()
+
+	err = realWAL.StoreCheckpoint(writer, tries...)
+	if err != nil {
+		return fmt.Errorf("error serializing checkpoint (%d): %w", checkpointNum, err)
+	}
+
+	duration := time.Since(startTime)
+	logger.Info().Float64("total_time_s", duration.Seconds()).Msgf("created checkpoint %d", checkpointNum)
+
+	return nil
+}
+
+func cleanupCheckpoints(checkpointer *realWAL.Checkpointer, checkpointsToKeep int) error {
+	// Don't list checkpoints if we keep them all
+	if checkpointsToKeep == 0 {
+		return nil
+	}
+	checkpoints, err := checkpointer.Checkpoints()
+	if err != nil {
+		return fmt.Errorf("cannot list checkpoints: %w", err)
+	}
+	if len(checkpoints) > int(checkpointsToKeep) {
+		// if condition guarantees this never fails
+		checkpointsToRemove := checkpoints[:len(checkpoints)-int(checkpointsToKeep)]
+
+		for _, checkpoint := range checkpointsToRemove {
+			err := checkpointer.RemoveCheckpoint(checkpoint)
+			if err != nil {
+				return fmt.Errorf("cannot remove checkpoint %d: %w", checkpoint, err)
+			}
+		}
+	}
+	return nil
+}
+
+// processTrieUpdate writes trie update to WAL, updates activeSegmentNum,
+// and gets tries from checkpointQueue for checkpointing if needed.
+// It also sends WAL update result and waits for trie update completion.
+func (c *Compactor) processTrieUpdate(
+	update *WALTrieUpdate,
+	checkpointQueue *CheckpointQueue,
+	activeSegmentNum int,
+	nextCheckpointNum int,
+) (
+	_activeSegmentNum int,
+	checkpointNum int,
+	checkpointTries []*trie.MTrie,
+) {
+
+	// RecordUpdate returns the segment number the record was written to.
+	// Returned segment number (>= 0) can be
+	// - the same as previous segment number (same segment), or
+	// - incremented by 1 from previous segment number (new segment)
+	segmentNum, skipped, updateErr := c.wal.RecordUpdate(update.Update)
+
+	// This ensures that updated trie matches WAL update.
+	defer func(updateResult error) {
+		// Send result of WAL update
+		update.ResultCh <- updateResult
+
+		// Wait for updated trie
+		trie := <-update.TrieCh
+		if trie == nil {
+			c.logger.Error().Msg("compactor failed to get updated trie")
+			return
+		}
+
+		checkpointQueue.Push(trie)
```

Comment:
⚠️  we should check `updateErr` before adding the trie to checkpoint queue, otherwise might cause inconsistency.

And if there is `updateError`, we probably don't have to wait for `<-update.TrieCh`, because we anyway won't add it to the checkpoint queue

Reply:
> :warning:  we should check updateErr before adding the trie to checkpoint queue, otherwise might cause inconsistency.

No, it is not needed.

`Ledger.Set()` checks `updateErr` and closes `TrieCh` channel without sending trie if `updateErr` is not nil.  When `Compactor` receives nil trie from `TrieCh`, `Compactor` logs it and nil trie isn't pushed to queue.

https://github.com/onflow/flow-go/blob/2127e44c0c86cf6af86ae967e6eba50ab99262df/ledger/complete/compactor.go#L343-L347

> And if there is updateError, we probably don't have to wait for <-update.TrieCh, because we anyway won't add it to the checkpoint queue

I think we should wait for `<-update.TrieCh` because it is `Ledger.Set` responsibility to send a trie matching WAL update.  And there is another situation that `Ledger.Set` would send a nil trie.  So I don't think we should special case it.

Reply:
Oh, I see. the trie channel is closed when fails to write WAL.